File: src/server/m3fs/m3fs.cc
/*
 * Copyright (C) 2015-2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/CmdArgs.h>
#include <base/Common.h>
#include <base/Errors.h>
#include <base/stream/IStringStream.h>

#include <m3/server/RequestHandler.h>
#include <m3/server/Server.h>
#include <m3/session/Disk.h>
#include <m3/session/M3FS.h>
#include <m3/session/ServerSession.h>
#include <m3/stream/Standard.h>

#include <limits>
#include <stdlib.h>

#include "backend/DiskBackend.h"
#include "backend/MemBackend.h"
#include "data/Dirs.h"
#include "data/INodes.h"
#include "sess/FileSession.h"
#include "sess/MetaSession.h"
#include "FSHandle.h"

using namespace m3;

class M3FSRequestHandler;

using base_class = RequestHandler<M3FSRequestHandler, M3FS::Operation, M3FS::COUNT, M3FSSession>;

static Server<M3FSRequestHandler> *srv;

class M3FSRequestHandler : public base_class {
public:
    explicit M3FSRequestHandler(WorkLoop *wl, Backend *backend, size_t extend, bool clear,
                                bool revoke_first, size_t max_load)
        : base_class(),
          _rgate(RecvGate::create(nextlog2<32 * M3FSSession::MSG_SIZE>::val,
                                  nextlog2<M3FSSession::MSG_SIZE>::val)),
          _handle(backend, extend, clear, revoke_first, max_load) {
        add_operation(M3FS::NEXT_IN, &M3FSRequestHandler::next_in);
        add_operation(M3FS::NEXT_OUT, &M3FSRequestHandler::next_out);
        add_operation(M3FS::COMMIT, &M3FSRequestHandler::commit);
        add_operation(M3FS::SYNC, &M3FSRequestHandler::sync);
        add_operation(M3FS::CLOSE, &M3FSRequestHandler::close_sess);
        add_operation(M3FS::FSTAT, &M3FSRequestHandler::fstat);
        add_operation(M3FS::SEEK, &M3FSRequestHandler::seek);
        add_operation(M3FS::STAT, &M3FSRequestHandler::stat);
        add_operation(M3FS::MKDIR, &M3FSRequestHandler::mkdir);
        add_operation(M3FS::RMDIR, &M3FSRequestHandler::rmdir);
        add_operation(M3FS::LINK, &M3FSRequestHandler::link);
        add_operation(M3FS::UNLINK, &M3FSRequestHandler::unlink);

        using std::placeholders::_1;
        _rgate.start(wl, std::bind(&M3FSRequestHandler::handle_message, this, _1));
    }

    virtual Errors::Code open(M3FSSession **sess, capsel_t srv_sel, const StringRef &args) override {
        size_t max_files = 64;
        if(args.length() > 0) {
            if(strncmp(args.c_str(), "files=", 6) == 0)
                max_files = IStringStream::read_from<size_t>(args.c_str() + 6);
        }
        *sess = new M3FSMetaSession(_handle, srv_sel, _rgate, max_files);
        return Errors::NONE;
    }

    virtual Errors::Code obtain(M3FSSession *sess, CapExchange &xchg) override {
        if(sess->type() == M3FSSession::META) {
            auto meta = static_cast<M3FSMetaSession *>(sess);
            if(xchg.in_args().length() == 0)
                return meta->get_sgate(xchg);
            return meta->open_file(srv->sel(), xchg);
        }
        else {
            auto file = static_cast<M3FSFileSession *>(sess);
            if(xchg.in_args().length() == 0)
                return file->clone(srv->sel(), xchg);
            return file->get_mem(xchg);
        }
    }

    virtual Errors::Code delegate(M3FSSession *sess, CapExchange &xchg) override {
        if(xchg.in_caps() != 1 || sess->type() != M3FSSession::FILE)
            return Errors::NOT_SUP;
        capsel_t sel = VPE::self().alloc_sel();
        static_cast<M3FSFileSession *>(sess)->set_ep(sel);
        xchg.out_caps(KIF::CapRngDesc(KIF::CapRngDesc::OBJ, sel, 1));
        return Errors::NONE;
    }

    virtual Errors::Code close(M3FSSession *sess) override {
        delete sess;
        _rgate.drop_msgs_with(ptr_to_label(sess));
        return Errors::NONE;
    }

    virtual void shutdown() override {
        _rgate.stop();
        _handle.flush_buffer();
        _handle.shutdown();
    }

    void next_in(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->next_in(is);
    }

    void next_out(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->next_out(is);
    }

    void commit(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->commit(is);
    }

    void sync(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->sync(is);
    }

    void seek(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->seek(is);
    }

    void fstat(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->fstat(is);
    }

    void stat(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->stat(is);
    }

    void mkdir(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->mkdir(is);
    }

    void rmdir(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->rmdir(is);
    }

    void link(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->link(is);
    }

    void unlink(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession *>();
        sess->unlink(is);
    }

    void close_sess(GateIStream &is) {
        M3FSSession *sess = is.label<M3FSSession*>();
        // reply first to prevent that we drop this message
        reply_vmsg(is, Errors::NONE);
        close(sess);
    }

private:
    RecvGate _rgate;
    FSHandle _handle;
};

NORETURN static void usage(const char *name) {
    cerr << "Usage: " << name
         << " [-n <name>] [-s <sel>] [-e <blocks>] [-c] [-r] [-b <blocks>]\n"
         << " [-o <offset>] (disk|mem <fssize>)\n";
    cerr << "  -n: the name of the service (m3fs by default)\n";
    cerr << "  -s: don't create service, use selectors <sel>..<sel+1>\n";
    cerr << "  -e: the number of blocks to extend files when appending\n";
    cerr << "  -c: clear allocated blocks\n";
    cerr << "  -r: revoke first, reply afterwards\n";
    cerr << "  -b: the maximum number of blocks loaded from the disk\n";
    cerr << "  -o: the file system offset in DRAM\n";
    exit(1);
}

int main(int argc, char *argv[]) {
    const char *name  = "m3fs";
    size_t extend     = 128;
    size_t max_load   = 128;
    bool clear        = false;
    bool revoke_first = false;
    goff_t fs_offset  = FS_IMG_OFFSET;

    int opt;
    while((opt = CmdArgs::get(argc, argv, "n:e:crb:o:")) != -1) {
        switch(opt) {
            case 'n': name = CmdArgs::arg; break;
            case 'e': extend = IStringStream::read_from<size_t>(CmdArgs::arg); break;
            case 'c': clear = true; break;
            case 'r': revoke_first = true; break;
            case 'b': max_load = IStringStream::read_from<size_t>(CmdArgs::arg); break;
            case 'o': fs_offset = IStringStream::read_from<goff_t>(CmdArgs::arg); break;
            default: usage(argv[0]);
        }
    }
    if(CmdArgs::ind >= argc)
        usage(argv[0]);

    WorkLoop wl;

    // create backend
    Backend *backend;
    const char *backend_type = argv[CmdArgs::ind];
    if(strcmp(backend_type, "disk") == 0)
        backend = new DiskBackend(&wl);
    else if(strcmp(backend_type, "mem") == 0) {
        if(CmdArgs::ind + 1 >= argc)
            usage(argv[0]);
        size_t fs_size = IStringStream::read_from<size_t>(argv[CmdArgs::ind + 1]);
        backend = new MemBackend(fs_offset, fs_size);
    }
    else
        usage(argv[0]);

    auto hdl = std::make_unique<M3FSRequestHandler>(&wl, backend, extend, clear, revoke_first, max_load);
    srv = new Server<M3FSRequestHandler>(name, &wl, std::move(hdl));

    wl.multithreaded(16);
    wl.run();

    delete backend;
    delete srv;
    return 0;
}

File: src/server/m3fs/src/m3fs.rs
/*
 * Copyright (C) 2020-2022 Nils Asmussen, Barkhausen Institut
 * Copyright (C) 2019-2020, Tendsin Mende <tendsin@protonmail.com>
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#![no_std]

#[macro_use]
extern crate m3;

mod backend;
mod buf;
mod data;
mod ops;
mod sess;

use crate::backend::{Backend, DiskBackend, MemBackend};
use crate::buf::{FileBuffer, MetaBuffer};
use crate::data::{Allocator, SuperBlock};
use crate::sess::{FSSession, M3FSSession, OpenFiles};

use m3::server::ExcType;
use m3::{
    boxed::Box,
    cap::Selector,
    cell::{LazyReadOnlyCell, LazyStaticRefCell, LazyStaticUnsafeCell, Ref, RefMut, StaticRefCell},
    col::{String, ToString, Vec},
    com::opcodes,
    env,
    errors::{Code, Error},
    io::LogFlags,
    server::{RequestHandler, Server, DEF_MAX_CLIENTS},
    tiles::OwnActivity,
};

// Server constants
const MSG_SIZE: usize = 128;

static SB: LazyStaticRefCell<SuperBlock> = LazyStaticRefCell::default();
// TODO we unfortunately need to use an unsafe cell here at the moment, because the meta buffer is
// basicalled used in all modules, making it really hard to use something like a RefCell here.
static MB: LazyStaticUnsafeCell<MetaBuffer> = LazyStaticUnsafeCell::default();
static FB: LazyStaticRefCell<FileBuffer> = LazyStaticRefCell::default();
static FILES: StaticRefCell<OpenFiles> = StaticRefCell::new(OpenFiles::new());
static BA: LazyStaticRefCell<Allocator> = LazyStaticRefCell::default();
static IA: LazyStaticRefCell<Allocator> = LazyStaticRefCell::default();
static SETTINGS: LazyReadOnlyCell<FsSettings> = LazyReadOnlyCell::default();
static BACKEND: LazyStaticRefCell<Box<dyn Backend>> = LazyStaticRefCell::default();

fn superblock() -> Ref<'static, SuperBlock> {
    SB.borrow()
}
fn superblock_mut() -> RefMut<'static, SuperBlock> {
    SB.borrow_mut()
}
fn meta_buffer_mut() -> &'static mut MetaBuffer {
    // safety: see comment for MB
    unsafe { MB.get_mut() }
}
fn file_buffer_mut() -> RefMut<'static, FileBuffer> {
    FB.borrow_mut()
}
fn open_files_mut() -> RefMut<'static, OpenFiles> {
    FILES.borrow_mut()
}
fn blocks_mut() -> RefMut<'static, Allocator> {
    BA.borrow_mut()
}
fn inodes_mut() -> RefMut<'static, Allocator> {
    IA.borrow_mut()
}
fn settings() -> &'static FsSettings {
    SETTINGS.get()
}
fn backend_mut() -> RefMut<'static, Box<dyn Backend>> {
    BACKEND.borrow_mut()
}

fn flush_buffer() -> Result<(), Error> {
    crate::meta_buffer_mut().flush()?;
    crate::file_buffer_mut().flush()?;

    // update superblock and write it back to disk/memory
    let mut sb = crate::superblock_mut();
    let inodes = crate::inodes_mut();
    sb.update_inodebm(inodes.free_count(), inodes.first_free());
    let blocks = crate::blocks_mut();
    sb.update_blockbm(blocks.free_count(), blocks.first_free());
    sb.checksum = sb.get_checksum();
    crate::backend_mut().store_sb(&sb)
}

#[derive(Clone, Debug)]
pub struct FsSettings {
    name: String,
    backend: String,
    mem_mod: String,
    extend: usize,
    max_load: usize,
    max_clients: usize,
    clear: bool,
    selector: Option<Selector>,
}

impl core::default::Default for FsSettings {
    fn default() -> Self {
        FsSettings {
            name: String::from("m3fs"),
            backend: String::from("mem"),
            mem_mod: String::from("fs"),
            extend: 128,
            max_load: 128,
            max_clients: DEF_MAX_CLIENTS,
            clear: false,
            selector: None,
        }
    }
}

fn usage() -> ! {
    println!(
        "Usage: {} [-n <name>] [-s <sel>] [-e <blocks>] [-c] [-f <name>] [-b <blocks>]",
        env::args().next().unwrap()
    );
    println!("       [-m <clients>] (disk|mem)");
    println!();
    println!("  -n: the name of the service (m3fs by default)");
    println!("  -s: don't create service, use selectors <sel>..<sel+1>");
    println!("  -e: the number of blocks to extend files when appending");
    println!("  -c: clear allocated blocks");
    println!("  -b: the maximum number of blocks loaded from the disk");
    println!("  -m: the maximum number of clients (receive slots)");
    println!("  -f: the name of the FS boot module ('fs' by default)");
    OwnActivity::exit_with(Code::InvArgs);
}

fn parse_args() -> Result<FsSettings, String> {
    let mut settings = FsSettings::default();

    let args: Vec<&str> = env::args().collect();
    let mut i = 1;
    while i < args.len() {
        match args[i] {
            "-n" => settings.name = args[i + 1].to_string(),
            "-f" => settings.mem_mod = args[i + 1].to_string(),
            "-s" => {
                if let Ok(s) = args[i + 1].parse::<Selector>() {
                    settings.selector = Some(s);
                }
            },
            "-e" => {
                settings.extend = args[i + 1]
                    .parse::<usize>()
                    .map_err(|_| String::from("Could not parse FS extend"))?;
            },
            "-b" => {
                settings.max_load = args[i + 1]
                    .parse::<usize>()
                    .map_err(|_| String::from("Could not parse max load"))?;
            },
            "-m" => {
                settings.max_clients = args[i + 1]
                    .parse::<usize>()
                    .map_err(|_| String::from("Failed to parse client count"))?;
            },
            "-c" => {
                settings.clear = true;
                i -= 1; // argument has no value
            },
            _ => break,
        }
        // move forward 2 by default, since most arguments have a value
        i += 2;
    }

    settings.backend = args[i].to_string();
    match settings.backend.as_str() {
        "mem" | "disk" => {},
        backend => return Err(format!("Unknown backend {}", backend)),
    }

    Ok(settings)
}

fn init_fs(mut backend: Box<dyn Backend>) {
    // init thread manager, otherwise the waiting within the file and meta buffer impl. panics.
    thread::init();

    let sb = backend.load_sb().expect("Unable to load super block");
    log!(LogFlags::FSInfo, "Loaded {:#?}", sb);

    BA.set(Allocator::new(
        String::from("Block"),
        sb.first_blockbm_block(),
        sb.first_free_block,
        sb.free_blocks,
        sb.total_blocks,
        sb.blockbm_blocks(),
        sb.block_size as usize,
    ));
    IA.set(Allocator::new(
        String::from("INodes"),
        sb.first_inodebm_block(),
        sb.first_free_inode,
        sb.free_inodes,
        sb.total_inodes,
        sb.inodebm_block(),
        sb.block_size as usize,
    ));

    // safety: we pass in a newly constructed MetaBuffer and have not initialized MB before
    unsafe {
        MB.set(MetaBuffer::new(sb.block_size as usize));
    }
    FB.set(FileBuffer::new(sb.block_size as usize));
    SB.set(sb);

    BACKEND.set(backend);
}

#[no_mangle]
pub fn main() -> Result<(), Error> {
    // parse arguments
    SETTINGS.set(parse_args().unwrap_or_else(|e| {
        println!("Invalid arguments: {}", e);
        usage();
    }));
    log!(LogFlags::FSInfo, "{:#?}", SETTINGS.get());

    // create and initialize backend for the file system
    let backend = if SETTINGS.get().backend == "mem" {
        Box::new(MemBackend::new(&SETTINGS.get().mem_mod)) as Box<dyn Backend>
    }
    else {
        Box::new(DiskBackend::new().expect("Failed to initialize disk backend!"))
            as Box<dyn Backend>
    };
    init_fs(backend);

    // create request handler and server
    // TODO just temporary: set a very high limit for client connections until we repair the way
    // the meta session is used.
    let mut hdl = RequestHandler::new_with(SETTINGS.get().max_clients, MSG_SIZE, 1024)
        .expect("Unable to create request handler");
    let mut srv =
        Server::new(&SETTINGS.get().name, &mut hdl).expect("Could not create service 'm3fs'");

    use opcodes::FileSystem;

    // register capability handler
    hdl.reg_cap_handler(FileSystem::Open, ExcType::Obt(2), FSSession::open);
    hdl.reg_cap_handler(FileSystem::GetMem, ExcType::Obt(1), FSSession::get_mem);
    hdl.reg_cap_handler(FileSystem::DelEP, ExcType::Del(1), FSSession::del_ep);
    hdl.reg_cap_handler(FileSystem::CloneFile, ExcType::Obt(2), FSSession::clone);
    hdl.reg_cap_handler(FileSystem::CloneMeta, ExcType::Obt(2), FSSession::clone);
    hdl.reg_cap_handler(FileSystem::SetDest, ExcType::Del(1), FSSession::set_dest);
    hdl.reg_cap_handler(
        FileSystem::EnableNotify,
        ExcType::Del(1),
        FSSession::enable_notify,
    );

    // register message handler
    hdl.reg_msg_handler(FileSystem::NextIn, FSSession::next_in);
    hdl.reg_msg_handler(FileSystem::NextOut, FSSession::next_out);
    hdl.reg_msg_handler(FileSystem::Commit, FSSession::commit);
    hdl.reg_msg_handler(FileSystem::Truncate, FSSession::truncate);
    hdl.reg_msg_handler(FileSystem::Close, FSSession::close);
    hdl.reg_msg_handler(FileSystem::FStat, FSSession::stat);
    hdl.reg_msg_handler(FileSystem::GetPath, FSSession::get_path);
    hdl.reg_msg_handler(FileSystem::Seek, FSSession::seek);
    hdl.reg_msg_handler(FileSystem::Sync, FSSession::sync);
    hdl.reg_msg_handler(FileSystem::Stat, FSSession::fstat);
    hdl.reg_msg_handler(FileSystem::Mkdir, FSSession::mkdir);
    hdl.reg_msg_handler(FileSystem::Rmdir, FSSession::rmdir);
    hdl.reg_msg_handler(FileSystem::Link, FSSession::link);
    hdl.reg_msg_handler(FileSystem::Unlink, FSSession::unlink);
    hdl.reg_msg_handler(FileSystem::Rename, FSSession::rename);
    hdl.reg_msg_handler(FileSystem::OpenPriv, FSSession::open_priv);

    hdl.run(&mut srv).expect("Server loop failed");

    Ok(())
}
